### PR TITLE
Publish uplink over AMQP

### PR DESCRIPTION
**Pull request by johanstokking**
And promoted types from `mqtt` to `core/types`

### CODE REVIEW

File: amqp/publisher.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+		err = action()
+	}
+	return err
+}
+
+func (p *DefaultPublisher) publish(key string, msg []byte, timestamp time.Time) error {
+	return p.do(func() error {
+		return p.channel.Publish(p.exchange, key, false, false, AMQP.Publishing{
+			ContentType:  "application/json",
+			DeliveryMode: AMQP.Persistent,
```
Comment (htdvisser) [inline amqp/publisher.go]: This is not consistent with https://github.com/TheThingsNetwork/ttn/issues/203

Comment (johanstokking) [inline amqp/publisher.go]: That's right, but #203 is more related to flaws in MQTT than with persistency in general

File: amqp/uplink.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/TheThingsNetwork/ttn/core/types"
+)
+
+// PublishUplink publishes an uplink message to the MQTT broker
```
Comment (htdvisser) [inline amqp/uplink.go]: AMQP

File: mqtt/downlink.go [hunk 3]
```diff
@@ -7,14 +7,16 @@ import (
 	"encoding/json"
 	"fmt"
 
+	"github.com/TheThingsNetwork/ttn/core/types"
+
```
Comment (htdvisser) [inline mqtt/downlink.go]: fmt

File: amqp/publisher.go (hunk 1 above)
Comment (htdvisser) [inline amqp/publisher.go]: 👍

**Pull request merged**